syntax = "proto3";

// The service for user interactions with the key-value store.
service KVS {
    // General operations
    rpc PutValue (PutArgs) returns (PutReply) {}
    rpc GetValue (GetArgs) returns (GetReply) {}
    rpc SetNX (LockArgs) returns (LockReply) {}
    rpc Del (LockArgs) returns (LockReply) {}
    rpc GetAllRows (GetArgs) returns (GetAllReply) {}
    rpc GetColsInRow (GetArgs) returns (GetAllReply) {}

    // Console operations
    rpc GetAllRowsByIp (GetArgs) returns (GetAllReply) {}
    rpc GetColsInRowByIp (GetArgs) returns (GetAllReply) {}
}

// The service types for server interactions with the key-value store.
enum OpType {
    PUT = 0;
    CPUT = 1;
    GET = 2;
    DELETE = 3;
    SETNX = 4;
    DEL = 5;
    GETALLROWS = 6;
    GETCOLSINROW = 7;
}

message Op {
    OpType Type = 1;
    string Row = 2;
    string Col = 3;
    string CurrValue = 4;
    string NewValue = 5;
    string RequestID = 6;
    string LockId = 7;
}

// A PutArgs is a message client sent to server for a put action.
message PutArgs {
    string Row = 1;
    string Col = 2;
    string CurrValue = 3;
    string NewValue = 4;
    int32 Option = 5;
    string RequestID = 6;
    string LockId = 7;
}

// A PutReply is a message server sent to client after a put action.
message PutReply {
    bool Success = 1;
}

// A GetArgs is a message client sent to server for a get action.
message GetArgs {
    string Row = 1;
    string Col = 2;
    string RequestID = 3;
    string LockId = 4;
}

// A GetReply is a message server sent to client after a get action.
message GetReply {
    bool Success = 1;
    string Value = 2;
}

// A DeleteArgs is a message client sent to server for a delete action.
message DeleteArgs {
    string Row = 1;
    string Col = 2;
    string RequestID = 3;
    string LockId = 4;
}

// A DeleteReply is a message server sent to client after a delete action.
message DeleteReply {
    bool Success = 1;
}

// A LockArgs is a message client sent to server for a setnx action.
message LockArgs {
    string Row = 1;
    int32 Timeout = 2;
    string LockId = 3;
    string RequestID = 4;
}

// A LockReply is a message server sent to client after a setnx action.
message LockReply {
    bool Success = 1;
}

// A GetAllRowsArgs is a message client sent to server for a get all rows action.
message GetAllRowsArgs {
}

// A GetAllRowsReply is a message server sent to client after a get all rows action.
message GetAllReply {
    repeated string item = 1;
}
